model/riscv_insts_zicbom.sail

```diff
@@ -0,0 +1,107 @@
+/*=======================================================================================*/
+/*  This Sail RISC-V architecture model, comprising all files and                        */
+/*  directories except where otherwise noted is subject the BSD                          */
+/*  two-clause license in the LICENSE file.                                              */
+/*                                                                                       */
+/*  SPDX-License-Identifier: BSD-2-Clause                                                */
+/*=======================================================================================*/
+
+// Cache Block Operations - Management
+//
+// Note that the CBOP prefetch instructions are existing base instructions that
+// are just defined to be hints, so there is not need to add any code for them
+// (since this model has no caches to prefetch into and the encodings already
+// exist).
+
+function cbo_clean_flush_enabled(p : Privilege) -> bool = feature_enabled_for_priv(p, menvcfg.CBCFE()[0], senvcfg.CBCFE()[0])
+function cbo_inval_enabled(p : Privilege) -> bool = feature_enabled_for_priv(p, menvcfg.CBIE()[0], senvcfg.CBIE()[0])
+function cbo_inval_as_inval(p : Privilege) -> bool = feature_enabled_for_priv(p, menvcfg.CBIE()[1], senvcfg.CBIE()[1])
+
+/* ****************************************************************** */
+union clause ast = RISCV_ZICBOM : (cbop_zicbom, regidx)
+
+mapping encdec_cbop : cbop_zicbom <-> bits(12) = {
+  CBO_CLEAN <-> 0b000000000001,
+  CBO_FLUSH <-> 0b000000000010,
+  CBO_INVAL <-> 0b000000000000
+}
+
+mapping clause encdec = RISCV_ZICBOM(cbop, rs1)             if haveZicbom()
+  <-> encdec_cbop(cbop) @ rs1 @ 0b010 @ 0b00000 @ 0b0001111 if haveZicbom()
+
+mapping cbop_mnemonic : cbop_zicbom <-> string = {
+  CBO_CLEAN <-> "cbo.clean",
+  CBO_FLUSH <-> "cbo.flush",
+  CBO_INVAL <-> "cbo.inval"
+}
+
+mapping clause assembly = RISCV_ZICBOM(cbop, rs1)
+  <-> cbop_mnemonic(cbop) ^ spc() ^ "(" ^ opt_spc() ^ reg_name(rs1) ^ opt_spc() ^ ")"
+
+val process_clean_inval : (regidx, cbop_zicbom) -> Retired
+function process_clean_inval(rs1, cbop) = {
+  let rs1_val = X(rs1);
+  let cache_block_size_exp = plat_cache_block_size_exp();
+  let cache_block_size = 2 ^ cache_block_size_exp;
+
+  // Offset to the beginning of the cache block.
+  let offset = (rs1_val & ~(zero_extend(ones(cache_block_size_exp)))) - rs1_val;
+
+  match ext_data_get_addr(rs1, offset, Read(Data), cache_block_size) {
```

Comment:
This probably needs a new hook for CHERI to allow "any byte in the cacheblock" rather than all bytes being covered.

E.g. extra_data_check_cmo_addr(rs1, Read(Data)) that just forwards to `ext_data_get_addr()` in the base RISC-V model?

Reply:
Good point, though I am not 100% sure on the intended CHERI behaviour - see https://github.com/riscv/riscv-cheri/issues/213

I added a `// TODO` at least so it hopefully doesn't get forgotten.